### PR TITLE
psh: add `date` command tests

### PR DESCRIPTION
**Pull request by niewim19**
JIRA: PD-249

## Description

Tests of `date` psh applet

## Motivation and Context

These tests were written alongside with `date` applet which is under review here:
 - https://github.com/phoenix-rtos/phoenix-rtos-utils/pull/123
This is a try of __Test Driven Developement__ procedure application for me.  

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## How Has This Been Tested?

- [ ] Already covered by automatic testing.
- [ ] New test added: (add PR link here).
- [ ] Tested by hand on: (list targets here).

## Checklist:

- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [ ] I have added tests to cover my changes.
- [ ] All new and existing linter checks and tests passed.
- [ ] My changes generate no new compilation warnings for any of the targets.

## Special treatment

- [ ] This PR needs additional PRs to work:
  - https://github.com/phoenix-rtos/phoenix-rtos-utils/pull/123
- [ ] I will merge this PR by myself when appropriate.

### CODE REVIEW

**Review by damianloew — COMMENTED** [review]
@niewim19 should I finish this test?

File: psh/test-date.py [hunk 1]
```diff
@@ -0,0 +1,156 @@
...
+    # Incorrect FORMAT passed when printing date
+    psh.assert_cmd(p, 'date +operand1', "operand1")
+    psh.assert_cmd(p, 'date', DFLT_DATE_REG, is_regex=True)
+
+    # too many arguments passed to print, should print first redundant
+    psh.assert_cmd(p, 'date operand1 operand2 operand3', unrec_arg("operand2"))
+    psh.assert_cmd(p, 'date', DFLT_DATE_REG, is_regex=True)
+
+    # nonexistent format '%k' passed to print
+    psh.assert_cmd(p, 'date +%Y%k%Y', "1970%k1970")
```
Comment (damianloew) [inline psh/test-date.py:98]: Please change 1970 to year regex.

**Review by damianloew — CHANGES_REQUESTED** [review]
Overall it would be nice to add this test. Please remember about specifying result argument for `psh.assert_cmd()`  (by default success) after rebase.

File: psh/test-date.py [hunk 2]
```diff
@@ -0,0 +1,154 @@
...
+# Author: Mateusz Niewiadomski, Mateusz Bloch
+#
+# This file is part of Phoenix-RTOS.
+#
+# %LICENSE%
+#
+
+import psh.tools.psh as psh
+
+DFLT_DATE_REG = r"\w{3}, \d{2} \w{3} \d{2} \d{2}:\d{2}:\d{2}\r+\n"  # default date when image is flashed
```
Comment (damianloew) [inline psh/test-date.py:17]: After consideration I think we should also pare such output (year can be printed using four digits, the comma after year and info about time zone are possible). Comment not valid now.

```suggestion
DFLT_DATE_REG = r"\w{3}, \d{2} \w{3} \d{2} \d{2}:\d{2}:\d{2}\r+\n"
```

File: psh/test-date.py [hunk 3]
```diff
@@ -0,0 +1,154 @@
...
+
+    psh.assert_cmd(p, "date -s @10000000000000000", expected=r"Sat, 25 Jan 55 \d{2}:\d{2}:\d{2}\r+\n", is_regex=True)
+    psh.assert_cmd(p, "date", expected=r"Sat, 25 Jan 55 \d{2}:\d{2}:\d{2}\r+\n", is_regex=True)
+    psh.assert_cmd(p, "date +%Y", expected="316889355")
+    psh.assert_cmd(p, "date -s @0", DFLT_DATE_REG, is_regex=True)
+
+    psh.assert_cmd(p, "date -d @10000000000000000", expected=r"Sat, 25 Jan 55 \d{2}:\d{2}:\d{2}\r+\n", is_regex=True)
+    psh.assert_cmd(p, "date", expected=DFLT_DATE_REG, is_regex=True)
+
+
```
Comment (damianloew) [inline psh/test-date.py:147]: Why checks like that were removed?
 - `psh.assert_cmd(p, "date -s @1600000000.1234567890", expected=SPEC_DATE_REG, is_regex=True)`

**Review by damianloew — CHANGES_REQUESTED** [review]

File: psh/test-date.py [hunk 4]
```diff
@@ -0,0 +1,197 @@
...
+    # Test help
+    psh.assert_cmd(p, "date -h", expected=r"Usage(.*?)FORMAT(.*?)\r+\n", is_regex=True)
+
+    # Test printing and formatting
+    psh.assert_cmd(p, "date", expected=DFLT_DATE_REG, is_regex=True)
+    psh.assert_cmd(p, "date +%Y", expected=r"\b(19|20)\d{2}\r+\n", is_regex=True)
+    psh.assert_cmd(p, "date +%H:%M:%Sformat", expected=r"\d{2}:\d{2}:\d{2}format\r+\n", is_regex=True)
+
+    # Test setting to value
+    psh.assert_cmd(p, "date -s @1600000000", expected=SPEC_DATE_REG, is_regex=True)
```
Comment (damianloew) [inline psh/test-date.py:49]: I think that we can expect `Sun, 13 Sep 20 12:26:40` (of course with optional comma and time zone info), I don't see the possibility where this value will differ, same for `-d`.

File: psh/test-date.py [hunk 5]
```diff
@@ -0,0 +1,197 @@
...
+    psh.assert_cmd(p, "date", expected=SPEC_DATE_REG, is_regex=True)
```
Comment (damianloew) [inline psh/test-date.py:50]: I'd leave only the following pattern (it will be flexible enough IMHO):
`12:26:4\d`

I assume it won't take more than 9s to get an output of the next `date` call.

File: psh/test-date.py [hunk 6]
```diff
@@ -0,0 +1,197 @@
...
+
+
+def test_edges(p):
+    """Test edge cases for epoch value passed to `date -s epoch"""
+
+    # maximum value for int32
+    psh.assert_cmd(
+        p,
+        "date -s @2147483647",
+        expected=r"Tue, 19 Jan (?:38|2038),? \d{2}:\d{2}:\d{2}(?:\w{3}|\w{4})?\r+\n",
```
Comment (damianloew) [inline psh/test-date.py:131]: Please place it in one variable and use exact value: `Tue, 19 Jan 38 03:14:07`, we can assume that time zones won't be added and an hour won't change (alternatively `GMT` will be printed).

File: psh/test-date.py [hunk 7]
```diff
@@ -0,0 +1,197 @@
...
+
+    # beyond int32 value
+    psh.assert_cmd(
+        p,
+        "date -s @2147483648",
+        expected=r"Tue, 19 Jan (?:38|2038),? \d{2}:\d{2}:\d{2}(?:\w{3}|\w{4})?\r+\n",
+        is_regex=True,
+    )
+    psh.assert_cmd(
+        p, "date", expected=r"Tue, 19 Jan (?:38|2038),? \d{2}:\d{2}:\d{2}(?:\w{3}|\w{4})?\r+\n", is_regex=True
```
Comment (damianloew) [inline psh/test-date.py:156]: There I'd suggest leaving some space and looking for `\d{2}` as seconds.

File: psh/test-date.py [hunk 8]
```diff
@@ -0,0 +1,197 @@
...
+    # too many arguments passed to print, should print first redundant
+    psh.assert_cmd(p, "date operand1 operand2 operand3", expected=unrec_arg("operand2"), result="fail")
+    psh.assert_cmd(p, "date", expected=DFLT_DATE_REG, is_regex=True)
+
+    # nonexistent format '%k' passed to print
+    psh.assert_cmd(p, "date +%Y%k%Y", expected=r"\b(19|20)\d{2}%k(19|20)\d{2}\r+\n", is_regex=True)
+    psh.assert_cmd(p, "date", expected=DFLT_DATE_REG, is_regex=True)
+
+
+def test_incorrect_datewrite(p):
```
Comment (damianloew) [inline psh/test-date.py:83]: It could be good to add a negative value for both `-s` and `-d`. Please remember that you can feel free to add new cases if you see the need.

**Review by damianloew — CHANGES_REQUESTED** [review]
Thanks for the last changes, some remarks and I'd merge it soon.

File: psh/test-date.py [hunk 9]
```diff
@@ -0,0 +1,189 @@
...
+#
+# %LICENSE%
+#
+
+import psh.tools.psh as psh
+
+DFLT_DATE_REG = r"\w{3}, \d{2} \w{3} (?:\d{2}|\d{4}),? \d{2}:\d{2}:\d{2}(?:\w{3}|\w{4})?\r+\n"
+SPEC_DATE_REG = r"Sun, 13 Sep (?:20|2020),? 12:26:4\d(?:\w{3}|\w{4})?\r+\n"  # arbitrary deta of epoch 1600000000 GMT
+HUGE_DATE_REG = r"Tue, 19 Jan (?:38|2038),? 03:14:0\d(?:\w{3}|\w{4})?\r+\n"  # arbitrary data of epoch 2147483647 GMT
+ZERO_DATE_REG = r"Thu, 01 Jan (?:70|1970),? 00:00:0\d(?:\w{3}|\w{4})?\r+\n"  # arbitrary deta of epoch 0 GMT
```
Comment (damianloew) [inline psh/test-date.py:20]:
```suggestion
ZERO_DATE_REG = r"Thu, 01 Jan (?:70|1970),? 00:00:0\d(?:\w{3}|\w{4})?\r+\n"  # arbitrary date of epoch 0 GMT
```

File: psh/test-date.py [hunk 10]
```diff
@@ -0,0 +1,189 @@
...
+# This file is part of Phoenix-RTOS.
+#
+# %LICENSE%
+#
+
+import psh.tools.psh as psh
+
+DFLT_DATE_REG = r"\w{3}, \d{2} \w{3} (?:\d{2}|\d{4}),? \d{2}:\d{2}:\d{2}(?:\w{3}|\w{4})?\r+\n"
+SPEC_DATE_REG = r"Sun, 13 Sep (?:20|2020),? 12:26:4\d(?:\w{3}|\w{4})?\r+\n"  # arbitrary deta of epoch 1600000000 GMT
+HUGE_DATE_REG = r"Tue, 19 Jan (?:38|2038),? 03:14:0\d(?:\w{3}|\w{4})?\r+\n"  # arbitrary data of epoch 2147483647 GMT
```
Comment (damianloew) [inline psh/test-date.py:19]:
```suggestion
HUGE_DATE_REG = r"Tue, 19 Jan (?:38|2038),? 03:14:0\d(?:\w{3}|\w{4})?\r+\n"  # arbitrary date of epoch 2147483647 GMT
```

**Review by damianloew — CHANGES_REQUESTED** [review]

**Review by damianloew — APPROVED** [review]
